File: onvif-gui/src/progress.cpp
/*******************************************************************************
* progress.cpp
*
* Copyright (c) 2020 Stephen Rhodes
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*
*******************************************************************************/

#include <iostream>
#include <QGridLayout>
#include <QFontMetrics>
#include <QMouseEvent>
#include <QPainter>
#include "progress.h"

Progress::Progress(QWidget* parent) : QWidget(parent)
{
    sldProgress = new ProgressSlider(Qt::Horizontal, this);

    lblProgress = new QLabel("0:00", this);
    setLabelWidth(lblProgress);

    lblDuration = new QLabel("0:00", this);
    setLabelWidth(lblDuration);

    lblPosition = new PositionLabel(this);
    setLabelHeight(lblPosition);

    QGridLayout* lytMain = new QGridLayout(this);
    lytMain->addWidget(lblPosition,   0, 1, 1, 1);
    lytMain->addWidget(lblProgress,   1, 0, 1, 1);
    lytMain->addWidget(sldProgress,   1, 1, 1, 1);
    lytMain->addWidget(lblDuration,   1, 2, 1, 1);
    lytMain->setColumnStretch(1, 10);
    lytMain->setContentsMargins(0, 0, 0, 0);
}

void Progress::setLabelWidth(QLabel* lbl) 
{
    lbl->setFixedWidth(lbl->fontMetrics().boundingRect("00:00:00").width());
    lbl->setAlignment(Qt::AlignCenter);
}

void Progress::setLabelHeight(QLabel* lbl) 
{
    lbl->setFixedHeight(lbl->fontMetrics().boundingRect("00:00:00").height());
}

QString Progress::getTimeString(qint64 milliseconds) const
{
    QString output;
    int position_in_seconds = milliseconds / 1000;
    int hours = position_in_seconds / 3600;
    int minutes = (position_in_seconds - (hours * 3600)) / 60;
    int seconds = (position_in_seconds - (hours * 3600) - (minutes * 60));
    char buf[32] = {0};
    if (hours > 0)
        sprintf(buf, "%02d:%02d:%02d", hours, minutes, seconds);
    else 
        sprintf(buf, "%d:%02d", minutes, seconds);

    output = QString(buf);
    return output;
}

void Progress::setDuration(qint64 arg) 
{
    if (arg > 0) {
        m_duration = arg;
        lblDuration->setText(getTimeString(arg));
    }
}

void Progress::setProgress(float pct) 
{
    sldProgress->setValue(sldProgress->maximum() * pct);
    lblProgress->setText(getTimeString((qint64)(m_duration * pct)));
}

ProgressSlider::ProgressSlider(Qt::Orientation orientation, QWidget* parent) : QSlider(orientation, parent)
{
    setMaximum(1000);
    setMouseTracking(true);
}

bool ProgressSlider::event(QEvent* event)
{
    if (event->type() == QEvent::Leave)
        ((Progress*)parent())->lblPosition->setText("", 0);
    return QSlider::event(event);
}

void ProgressSlider::mouseMoveEvent(QMouseEvent* event)
{
    Progress* p = (Progress*)parent();
    qint64 duration = p->duration();
    int x = event->pos().x();
    if (duration) {
        double percentage = x / (double)width();
        double position = percentage * duration;
        p->lblPosition->setText(p->getTimeString((int)position), x);
    }
}

void ProgressSlider::mousePressEvent(QMouseEvent* event)
{
    float pct = event->pos().x() / (float)width();
    ((Progress*)parent())->emit seek(pct);
}

PositionLabel::PositionLabel(QWidget* parent) : QLabel(parent)
{

}

void PositionLabel::setText(const QString& arg, int pos)
{
    QLabel::setText(arg);
    m_pos = pos;
}

void PositionLabel::paintEvent(QPaintEvent *event)
{
    QPainter painter;
    painter.begin(this);
    QFontMetrics metrics = fontMetrics();
    QRect rect = metrics.boundingRect(text());
    int x = std::min(width() - rect.width(), m_pos);
    painter.drawText(QPoint(x, height()), text());
    painter.end();
}

File: onvif-gui/onvif_gui/components/progress.py
#********************************************************************
# libonvif/onvif-gui/onvif_gui/components/progress.py
#
# Copyright (c) 2023  Stephen Rhodes
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#*********************************************************************/

from PyQt6.QtWidgets import QSlider, QLabel, QWidget, QGridLayout
from PyQt6.QtCore import Qt, QPoint
from PyQt6.QtGui import QPainter

class Slider(QSlider):
    def __init__(self, o, P):
        super().__init__(o)
        self.P = P
        if P.showPosition:
            self.setMouseTracking(True)
        self.setFocusPolicy(Qt.FocusPolicy.NoFocus)

    def leaveEvent(self, e):
        self.P.updatePosition(-1, 0)

    def mousePressEvent(self, e):
        pct = e.position().x() / self.width()
        uri = self.P.mw.glWidget.focused_uri
        player = self.P.mw.pm.getPlayer(uri)
        if player is not None:
            player.seek(pct)

    def mouseMoveEvent(self, e):
        x = e.position().x()
        pct = x / self.width()
        self.P.updatePosition(pct, x)

class Position(QLabel):
    def __init__(self):
        super().__init__()
        self.pos = 0

    def setText(self, s, n):
        self.pos = n
        super().setText(s)

    def paintEvent(self, e):
        painter = QPainter(self)
        rect = self.fontMetrics().boundingRect(self.text())
        x = min(self.width() - rect.width(), self.pos)
        painter.drawText(QPoint(int(x), self.height()), self.text())

class Progress(QWidget):
    def __init__(self, mw):
        super().__init__()
        self.mw = mw
        self.duration = 0
        self.showPosition = True

        self.sldProgress = Slider(Qt.Orientation.Horizontal, self)
        self.sldProgress.setMaximum(1000)
        self.lblProgress = QLabel("0:00")
        self.setLabelWidth(self.lblProgress)
        self.lblDuration = QLabel("0:00")
        self.setLabelWidth(self.lblDuration)
        if self.showPosition:
            self.lblPosition = Position()

        lytProgress = QGridLayout(self)
        if self.showPosition:
            lytProgress.addWidget(self.lblPosition,  0, 1, 1, 1)
        lytProgress.addWidget(self.lblProgress,      1, 0, 1, 1)
        lytProgress.addWidget(self.sldProgress,      1, 1, 1, 1)
        lytProgress.addWidget(self.lblDuration,      1, 2, 1, 1)
        lytProgress.setContentsMargins(0, 0, 0, 0)
        lytProgress.setColumnStretch(1, 10)

    def setLabelWidth(self, l):
        l.setFixedWidth(l.fontMetrics().boundingRect("00:00:00").width())
        l.setAlignment(Qt.AlignmentFlag.AlignCenter)

    def updateDuration(self, n):
        self.duration = n
        self.lblDuration.setText(self.timestring(n))

    def updateProgress(self, f):
        self.sldProgress.setValue(int(f * 1000))
        self.lblProgress.setText(self.timestring(int(self.duration * f)))

    def updatePosition(self, f, n):
        if self.showPosition:
            if f >= 0:
                position = int(f * self.duration)
                self.lblPosition.setText(self.timestring(position), n)
            else:
                self.lblPosition.setText("", 0)

    def timestring(self, n):
        time_interval = int(n / 1000)
        hours = int(time_interval / 3600)
        minutes = int ((time_interval - (hours * 3600)) / 60)
        seconds = int ((time_interval - (hours * 3600) - (minutes * 60)))
        if hours > 0:
            buf = "%02d:%02d:%02d" % (hours, minutes, seconds)
        else:
            buf = "%d:%02d" % (minutes, seconds)
        return buf

